packages/story-editor/src/app/rightClickMenu/menus/emptyStateMenu.js

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { RIGHT_CLICK_MENU_LABELS } from '../constants';
+import { states, useHighlights } from '../../highlights';
+import { MediaUploadButton } from '../../../components/form';
+import LibraryProvider from '../../../components/library/libraryProvider';
+import useOnMediaSelect from '../../../components/library/panes/media/local/useOnMediaSelect';
+
+const MediaButton = () => {
+  const { onSelect } = useOnMediaSelect();
+
+  return (
+    <MediaUploadButton
+      renderButton={(open) => (
+        // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
+        <span onClick={open}>
```

Comment:
Hm, why not use an actual button?

Reply:
That would be a button in button with extra styling and clickable area changes needed:

![image](https://user-images.githubusercontent.com/21036224/150535203-b1193008-4bbc-4ad0-a3db-0946272b84d7.png)

Reply:
Ah, okay. Maybe leave an additional comment to `eslint-disable-next-line`, e.g. `// Disable reason: the parent menu item is a button`.

Alternatively... the menu items don't really _have to_ be buttons themselves, e.g. see the example https://google.github.io/web-stories-wp/storybook/?path=/story/designsystem-components-contextmenu--default with the text `i am neither a button nor a link`. So one option could be rendering this media upload button as the menu item directly instead of having it passed in as a label. Would need some adjustments to rendering the right-click menu since currently everything is rendered as buttons by default, but might be a better option. WDYT?

Reply:
I think it's a one-off case so I would stick with it until we need more flexibility in the context menu (non-trivial implementation with some use-cases to cover).
Reason added 👍🏻

Reply:
The ideal solution would be to have the media upload button **be** the actual menu item button. But I assume that's not really feasible here?

Reply:
> But I assume that's not really feasible here?

It is, that's what I was suggesting as well.

We could basically assign the menu item some kind of param (e.g. type: 'mediaUpload') in the app folder when defining menu items, and then, when rendering the items in https://github.com/google/web-stories-wp/blob/417c819d3c12adaf5e33a6c29f59c2050fabeaeb/packages/story-editor/src/components/canvas/rightClickMenu.js#L97 render the Media Upload button instead of the `ContextMenuComponents.MenuButton`.

Would avoid having rendering a button as a `span` and a label.

Reply:
@merapi Hm, couldn't it be something like

```
...
{rightClickMenuItems.map(
              ({
                label,
                separator,
                onSelect
                ...buttonProps
              }) => (
...
<MediaUploadButton
          renderButton={<RenderButton ...buttonProps label={label}}
          onInsert={onSelect}
     />
...
```
?
where `onSelect` could come from `useEmptyStateMenu`.
In the example above, it looks like you're wrapping the `MediaUpload` around label etc., however, all those props should go to the `renderButton` directly.

Thoughts?

Reply:
Sure, maybe I'm thinking too much about _separation of concerns_ (polluting the RightClickMenu with MediaUploadButton specific code), I can do it both ways, no hard feelings.

Let's vote? 😄 
- [ ] @merapi 
- [ ] @miina 
- [ ] @barklund

Reply:
I agree with @miina, that the current solution is pretty hacky (and potentially inaccessible). But I also agree with @merapi, that I don't want any media button logic inside the generic menu item button.
 
 Could we perhaps create some form of callback, that the menu can invoke before rendering the individual button, which we could use for this button, to "enrich" it with the media upload logic?

Something like:

```js
rightClickMenuItems.map(
  ({
    label,
    separator,
    onSelect,
    enrich = null,
    ...buttonProps
  }) => {
  const button = <...>;
  return enrich ? enrich(button) : button;
})
```

Then we could use this enrich function to override the `onClick` and add the extra logic.

The logic from the media upload button would then be moved to a custom hook that could then be used in that button and this button.

It's quite a bit of work, but I think it's worth the effort.

Reply:
> that I don't want any media button logic inside the generic menu item button.

I also don't want media button logic inside the generic menu item button, but this wouldn't influence the generic menu button item (`MenuButton` inside the design-system), it would just influence the specific usage of the Right Click menu components based on what's needed when rendering.

The `MediaButton` would be displayed **instead** of the generic item button, not inside it, e.g. something like
If it's media upload button, render that, otherwise render the generic menu item button.
```
...
{rightClickMenuItems.map(
              ({...}) => (
...
{isMediaUpload ? <MediaUploadButton
          renderButton={<RenderButton ...buttonProps label={label}}
          onInsert={onSelect}
     /> : <MenuButton> ... </MenuButton>}
...
```

Or perhaps I understood incorrectly what you meant by "media button logic inside the generic menu item button".

Worth noting that this part will also change in https://github.com/GoogleForCreators/web-stories-wp/pull/10199/files#diff-72ed6cad8c593916cce1ece6488790bff7059731e2a95030152549d3712711ffR104-R110 where we'll render `SubMenuTrigger` instead of typical `MenuButton` in case the menu item has submenu -- would be good to take it into account when making changes here.

Reply:
My _SoC_ was about RightClickMenu doing its own thing without having to know something like MediaUploadButton even exists (@barklund also?)

Should we create a follow up ticket with one of the solutions, because it will need its own round of reviews/testing?

BTW: if we  are talking about hacks, it all could be done with:
```js
{
      label: RIGHT_CLICK_MENU_LABELS.UPLOAD_IMAGE_OR_VIDEO,
      onClick: () => {
        setHighlights({
          highlight: states.MEDIA,
        });
        document.querySelector(`[aria-label="Upload"]`).click();
      },
      ...menuItemProps,
    },
```

Reply:
> My SoC was about RightClickMenu doing its own thing without having to know something like MediaUploadButton even exists

Alright, it seemed to me that since `RightClickMenu` component inside the canvas folder is about just rendering anyway then it's ok to "know" about `MediaUpload`, it makes sense to me that it can support different types of items, even if currently, it's used just once.

Anyway -- ok by me to follow up with this separately.

@BrittanyIRL and / @samwhale might also have some suggestions here about rendering a custom button type (media upload button) in `canvas/rightClickMenu.js` since AFAIK they've mostly created the context menu logic.

Reply:
> My SoC was about RightClickMenu doing its own thing without having to know something like MediaUploadButton even exists

One more thought: it seems to me that `app/rightClickMenu/menus/emptyStateMenu.js` shouldn't actually have to "know" about `MediaUpload` -- it would ideally be just the menu item definitions and quite simple labels without the items having to be aware how they are rendered.

Maybe let's wait until EOD today (or tomorrow European morning) to see if @barklund (or anyone else) has any new thoughts and otherwise just choose an option that's correct in terms of accessibility, and go with it, and create a follow-up code quality issue. Thoughts?

Reply:
Yeah, let's go with a semantically correct solution for now even if at the loss of code quality. And then create a follow-up for code quality for proper refactor.

Reply:
💭 The media upload component has been a thorn in the side of the right click menu for quite some time. This is what sparked the refactor of the right click menu.

The refactor focused on making the context menu more composable so that individual elements could be styled/wrapped when necessary. Previous to the refactor we did [this hacky thing](https://github.com/GoogleForCreators/web-stories-wp/pull/9960/files#diff-c62ae1c4616a0518137fda6727273417d4c0b4692e40e9cc7a7bdbd519713850L354-L371)

---

The tricky part here is that we're mapping through an array of objects and generating menu items based on the properties of the objects.

The goal is not to wrap every button with the `MediaUpload` component if it's not necessary.

>My SoC was about RightClickMenu doing its own thing without having to know something like MediaUploadButton even exists

This would be my goal too. Unfortunately the `MediaUpload` component uses a render prop to expose the `open` function to its children. So imo the best place to add this would be in the provider. If a property is added to menus that need this special button, then the generator can wrap those buttons specifically with the `MediaUpload` component.

```js
...
{rightClickMenuItems.map(
              ({ isMediaUpload, onClick, ...rest }) => (
...
{isMediaUpload ? <MediaUploadButton
          renderButton={(open) => <ContextMenuComponents.Button onClick={(evt) => { onClick(evt); open(evt); }} {...buttonProps} {...rest} />}
          onInsert={onSelect}
     /> : <ContextMenuComponents.Button> ... </ContextMenuComponents.Button>}
...
```

I'd go with Miina's approach here. This way the MediaUpload button can be rendered only when necessary.

Reply:
Let's continue in https://github.com/GoogleForCreators/web-stories-wp/issues/10393

Reply:
@merapi As far as I understand from @samwhale's comment then refactoring to allow rendering custom items was already done exactly for that and the suggested way would be conditionally rendering the `MediaUploadButton` if `isMediaUpload` as in Sam's comment. To be confirmed by Samuel but not sure if there's anything else to improve within #10393 if https://github.com/GoogleForCreators/web-stories-wp/pull/9960 was already specifically for that. 🤔

Reply:
I just realized that we also have the `pageSideMenu` that's also using the context menu and has the Media Upload item inside it and it's done the same way as suggested in Sam's comment:
https://github.com/GoogleForCreators/web-stories-wp/blob/1bc0224f2b28f303918d8175d24f405b552e6fe2/packages/story-editor/src/components/canvas/pageSideMenu.js#L127-L138

Even if we need to follow up on this and there are more things to improve, then for consistency, would be nice to do it the same way until then.

Reply:
> @merapi As far as I understand from @samwhale's comment the refactoring to allow rendering custom items was already done exactly for that and the suggested way would be conditionally rendering the `MediaUploadButton` if `isMediaUpload` as in Sam's comment

Yes exactly 👍

Reply:
> To be confirmed by Samuel but not sure if there's anything else to improve within #10393 if #9960 was already specifically for that. 🤔

I'm also confused about why a new issue is needed.

---

The ticket description:

> We want to render custom items like MediaUploadButton, without the parent button element that comes from RightClickMenu.

The button element in the right click menu would be the child of `MediaUploadButton`, no? The `MediaUpload` component renders the UI passed in the `renderButton` prop. 

I guess since `MediaUpload` is being passed in through config -> there's no guarantee that no other UI will be rendered (in the case of that component being swapped out with something else). Is this what you're referring to?

Reply:
@miina @samwhale PTAL at [this](https://github.com/GoogleForCreators/web-stories-wp/pull/10261/commits/4dfc52619f290cb297f27f213a54a9f8f36adec9) `enrich` approach inspired by @barklund - looks pretty clean (not polluting the RCM) and powerful (you can do whatever you want) to me, WDYT?

Reply:
@merapi Looks fine, however, the main thing is that it seems to me that the `app` folder should not be polluted with rendering -- it's the RCM file (`...canvas/rightClickMenu.js`) that's meant for rendering so it's okay to have custom rendering there.

The `app/rightClickMenu/menus/emptyStateMenu.js` is just for defining the menu items and it shouldn't really concern itself a lot with _how_ these get rendered -- e.g. if these same menu items would be rendered in a completely different way or place, the content of the `app/rightClickMenu` should ideally not have to change since it's just defining the items but not rendering. 

So to me it seems exactly the opposite -- it's ok to customize the RCM component inside the `canvas` folder since that's specifically meant for rendering based on how it's needed in the editor per UX -- that's why this file exists, to render the menu in a custom way specific to the editor. And we should not pollute what's inside `app/rightClickMenu` folder since the item definitions should not be concerned with how they are rendered. So I don't think the `app/rightClickMenu/menus/emptyStateMenu.js` should be aware of `MediaUploadButton` or `LibraryProvider` and be responsible of rendering those. 🤷 

Anyway -- if you and @samwhale and @barklund think otherwise then it's ok with me as well.

Reply:
> So I don't think the app/rightClickMenu/menus/emptyStateMenu.js should be aware of MediaUploadButton or LibraryProvider and be responsible of rendering those. 🤷

Personally, I think `emptyStateMenu` should be the only one concerned about those because it's the only menu that requires them.

> So to me it seems exactly the opposite -- it's ok to customize the RCM component inside the canvas folder since that's specifically meant for rendering based on how it's needed in the editor per UX -- that's why this file exists, to render the menu in a custom way specific to the editor.

It's in `canvas` folder, but for me, it's mainly a component (`components/canvas/rightClickMenu.js`) that can be moved to any other project and reused there (in theory), so tying that with `MediaUploadButton` just prevents that.

I was concerned that `isMediaUpload` solution will lead us to the RCM component that has `isMediaUpload, isAnotherButtonThatRequiresProvider` and so on. But if it makes more sense for the majority - I can do it.

Reply:
> Personally, I think `emptyStateMenu` should be the only one concerned about those because it's the only menu that requires them.

Actually, the page menu also requires the media upload button (see the `PageSideMenu` component inside the `canvas` folder). It's just not using `MediaUploadButton` wrapper (using custom `MediaPicker` wrapper instead) but could use that (since`MediaUploadButton` is used in most of the places).

So in both of those menus we could instead use `isMediaPicker` or `wrapWithMediaPicker` (that's currently used in `PageSideMenu`. Whatever we do, it would be good to unify the approach in these two menus for consistency.

Reply:
Good morning everybody.

Tough problem due to the library provider being nested inside the right click provider. I'm hoping not to add another instance of the library provider just for this button. 

https://github.com/GoogleForCreators/web-stories-wp/blob/4dfc52619f290cb297f27f213a54a9f8f36adec9/packages/story-editor/src/app/rightClickMenu/menus/emptyStateMenu.js#L48-L52

The library provider doesn't depend on the right click provider, so what do you think about raising the current library provider up above the right click provider? I think that would be ideal.

---

The `enrich` approach works here, but there is one caveat: any `onClick` that is passed to the `MenuItem` will be overridden by the `cloneElement` call.

https://github.com/GoogleForCreators/web-stories-wp/blob/4dfc52619f290cb297f27f213a54a9f8f36adec9/packages/story-editor/src/components/canvas/rightClickMenu.js#L102-L104

https://github.com/GoogleForCreators/web-stories-wp/blob/4dfc52619f290cb297f27f213a54a9f8f36adec9/packages/story-editor/src/app/rightClickMenu/menus/emptyStateMenu.js#L37-L39

The `MediaUpload` button is a special case where no `onClick` function is being passed into the component, so that won't be a problem here.

However - I think the `enrich` method is more difficult to understand than using a ternary like suggested here: https://github.com/GoogleForCreators/web-stories-wp/pull/10261#discussion_r790834837. Adding the `enrich` function where you have to pass a react component as a prop is a little strange.

If we wanted to keep this approach I'd prefer to pass in the button as a child, rather than as a prop. I'd reserve passing in component references as props and react elements as children.

However, for maintainability sake I think it'd be better to not add this abstraction. Thoughts?

Reply:
> The enrich approach works here, but there is one caveat: any onClick that is passed to the MenuItem will be overridden by the cloneElement call.

I think you would either use simple `onClick` defined in the MenuItem or use `enrich` for something more complex.
I don't see any use case, but you can just do:
```js
onClick: (e) => {
  OriginalButton.props.onClick(e);
  open(e);
},
```
---
> Adding the enrich function where you have to pass a react component as a prop is a little strange.

As a param. It's kinda like [HOC](https://reactjs.org/docs/higher-order-components.html).
`Concretely, a higher-order component is a function that takes a component and returns a new component.`

---

> So in both of those menus we could instead use isMediaPicker or wrapWithMediaPicker (that's currently used in PageSideMenu. Whatever we do, it would be good to unify the approach in these two menus for consistency.

Those two actions (in sidemenu & emptystate) do different things, and the internal `insertMediaElement` is different (first replaces the media, second just inserts it).
Do we still want to unify them? or just create this `isMediaUpload` prop that handles the empty state menu upload case.

Reply:
> Those two actions (in sidemenu & emptystate) do different things, and the internal insertMediaElement is different (first replaces the media, second just inserts it).
Do we still want to unify them? or just create this isMediaUpload prop that handles the empty state menu upload case.

Ok by me to just create `isMediaUpload` for now!
In ideal case, `PageSideMenu` should also use `MediaUpoadButton` where `onSelect` (replacing background) would come from the menu item definition. At the first glance, it looks like the `MediaPicker` is very similar to `MediaUploadButton` and could be just replaced and removed. We can just create a follow-up code quality issue to do that separately, no need to look into it now.

Reply:
> > Adding the enrich function where you have to pass a react component as a prop is a little strange.
> 
> As a param. It's kinda like [HOC](https://reactjs.org/docs/higher-order-components.html). `Concretely, a higher-order component is a function that takes a component and returns a new component.`

Right, I understand. Let me rephrase: I'm having trouble understanding the benefits of using an HOC pattern here. Couldn't we get the same functionality without this abstraction?

>[The purpose of a HOC is to enhance a component (usually a dumb component) with extra functionality](https://levelup.gitconnected.com/real-world-examples-of-higher-order-components-hoc-for-react-871f0d8b39d8)

The HOC pattern is primarily implemented to re-use component logic that's baked into a component. However, there's nothing that needs to be "re-used" here. This is because the `MediaUpload` component exposes functionality through its render prop.

I can see the idea of wanting to inject new functionality (`onClick` from `MediaUpload`) but I think I'd prefer this:

```jsx
<ContextMenu>
    ...
    <MediaUpload
        render={(open) => {
            <ContextMenuComponents.MenuItem 
               onClick={(evt) => {
                  open(evt);
                  onClick(evt);
               }}
               {...props}
             >
                 {label}
            </ContextMenuComponents.MenuItem>
        }
    />
</ContextMenu>
```

---

I think this conversation is highlighting the inflexibility of the functionality we need to access through the `MediaUpload` component.

The `MediaUpload` component is passed into the application through the configuration. The `MediaUpload` component functions like context, giving children access to one function. IMO the `MediaUpload` component should have been a function because it is not rendering any UI. This would give us more flexibility to use it in scenarios like this.

However, I know this is an API that has been around for a while, and probably for good reason too. The best path forward would be to keep this as maintainable as possible. If you believe HOC is the most maintainable then that works  for me as well.

Reply:
>it looks like the `MediaPicker` is very similar to `MediaUploadButton` and could be just replaced and removed

The `MediaPicker` was created before the context menu was refactored, and was the best way we had to access the `open` function from the `MediaUpload` component.

Now that the context menu is refactored - the `MediaPicker` component should be removed and we should do something like the above comment.

Reply:
> Ok by me to just create isMediaUpload for now!
In ideal case, PageSideMenu should also use MediaUpoadButton where onSelect (replacing background) would come from the menu item definition. At the first glance, it looks like the MediaPicker is very similar to MediaUploadButton and could be just replaced and removed. We can just create a follow-up code quality issue to do that separately, no need to look into it now.

Okay, right now we have a fully working & accessible solution that at least half of us agree on.
It makes sense to try to unify the `PageSideMenu` part, so I've updated the #10393 accordingly, for now - let's unblock this ticket.